### PR TITLE
ck tile pagedkv prefill

### PR DESCRIPTION
**Pull request by ltqin**

### CODE REVIEW

File: include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        });
+
+        // __builtin_amdgcn_sched_group_barrier(0x100, 2, 0); // DS read
+        // __builtin_amdgcn_sched_group_barrier(0x008, 2, 0); // MFMA
+        // __builtin_amdgcn_sched_group_barrier(0x100, 2, 0); // DS read
+        // __builtin_amdgcn_sched_group_barrier(0x008, 2, 0); // MFMA
+        // __builtin_amdgcn_sched_group_barrier(0x100, 2, 0); // DS read
+        // __builtin_amdgcn_sched_group_barrier(0x008, 2, 0); // MFMA
+        // __builtin_amdgcn_sched_group_barrier(0x100, 2, 0); // DS read
+        // __builtin_amdgcn_sched_group_barrier(0x008, 10, 0); // MFMA
```
Comment (poyenc) [inline include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp]: please remove no-longer used code

Comment (ltqin) [reply include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp]: done

File: example/ck_tile/01_fmha/fmha_fwd.hpp [hunk 2]
```diff
@@ -178,6 +178,86 @@ struct fmha_fwd_args
...
+    ck_tile::index_t hdim_q;
+    ck_tile::index_t hdim_v;
+    ck_tile::index_t nhead_q;
+    ck_tile::index_t nhead_k;
+
+    float scale_s;
+    float scale_p;
+    float scale_o;
+
+    float logits_soft_cap;
```
Comment (poyenc) [inline example/ck_tile/01_fmha/fmha_fwd.hpp:235]: do we really need to support logits soft-capping in this kernel?

File: example/ck_tile/01_fmha/fmha_fwd.cpp [hunk 3]
```diff
@@ -1082,6 +1099,18 @@ bool run(const ck_tile::ArgParser& arg_parser)
 
             return fmha_fwd_splitkv(fmha_splitkv_traits, fmha_splitkv_args, stream_config);
         }
+#endif
+#if CK_TILE_FMHA_FWD_PAGEDKV_API
+        if(use_kvcache)
```
Comment (poyenc) [inline example/ck_tile/01_fmha/fmha_fwd.cpp:1104]: If we enable both CK_TILE_FMHA_FWD_SPLITKV_API & CK_TILE_FMHA_FWD_PAGEDKV_API. We will call both of `fmha_fwd_splitkv()` and `fmha_fwd_pagedkv()` if `1 < num_splits && use_kvcache` is **true**

File: include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        a_block_tensor.get_thread_buffer() = a_block_tensor_tmp.get_thread_buffer();
+
+        // construct B-warp-window
+        auto b_warp_window_tmp = make_tile_window(
+            b_block_window_tmp.get_bottom_tensor_view(),
+            make_tuple(number<WG::kN>{}, number<WG::kK>{}),
+            b_block_window_tmp.get_window_origin() + multi_index<2>{iNWarp * WG::kN, 0},
+            make_static_tile_distribution(typename WG::BWarpDstrEncoding{}));
+
+#if 0 // FIXME: using array will cause register spill
```
Comment (poyenc) [inline include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp]: can we just remove the code in `#if 0` block?

File: include/ck_tile/ops/fmha/pipeline/tile_fmha_traits.hpp [hunk 5]
```diff
@@ -37,6 +37,34 @@ struct TileFmhaTraits
...
+          bool kPadHeadDimV_ /* paddding for hdim_v */,
+          bool kHasLogitsSoftCap_,
+          BlockAttentionBiasEnum BiasEnum_,
+          bool kHasBiasGrad_,
+          bool kStoreLSE_, /* set to true if either num_splits > 1 or fwd training is running */
+          bool kIsPagedKV_,
+          bool kDoFp8StaticQuant_,
+          index_t kBlockPerCu_  = -1, /* overwrite occupancy if not -1 */
+          bool kSkipMinSeqlenQ_ = false /* skip min seqlen q while chunked prefill */>
+struct TileFmhaFwdPagedKVTraits
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/tile_fmha_traits.hpp:52]: Can we just let `TileFmhaFwdPagedKVTraits<>` to inherit from `TileFmhaTraits<>`?

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp [hunk 6]
```diff
@@ -61,6 +61,59 @@ struct BlockFmhaPipelineProblem
...
+          typename LSEDataType_,
+          typename PDataType_,
+          typename OaccDataType_,
+          typename ODataType_,
+          typename BlockFmhaShape_,
+          bool kIsGroupMode_,
+          typename AttentionVariant_,
+          typename FmhaMask_,
+          typename Traits_>
+struct BlockFmhaFwdPagedKVPipelineProblem
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp:79]: Can we just let `BlockFmhaFwdPagedKVPipelineProblem<>` to inherit from `BlockFmhaPipelineProblem<>`?

File: include/ck_tile/ops/fmha/pipeline/block_fmha_fwd_pagedkv_pipeline_qr_ks_vs.hpp [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+#pragma once
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/fmha/block/block_attention_bias_enum.hpp"
+#include "ck_tile/ops/fmha/pipeline/block_fmha_fwd_pagedkv_pipeline_qr_ks_vs_default_policy.hpp"
+#include "ck_tile/ops/reduce/block/block_reduce.hpp"
+
+namespace ck_tile {
+
+// This pipeline is qkv all located in LDS
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_fwd_pagedkv_pipeline_qr_ks_vs.hpp]: Add a **TODO** comment indicating that this class is a variant of the existing `BlockFmhaFwdSplitKVPipelineQRKSVS` pipeline. Refactoring to extract shared logic is recommended as future work.

File: include/ck_tile/ops/fmha/kernel/fmha_fwd_pagedkv_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,1372 @@
...
+// S[seqlen_q, seqlen_k] = Q[seqlen_q, hdim_q] @ K[seqlen_k, hdim_q]
+// S'[seqlen_q, seqlen_k] = S[seqlen_q, seqlen_k] * Scale[1]
+// S''[seqlen_q, seqlen_k] = S'[seqlen_q, seqlen_k] + Bias[seqlen_q, seqlen_k]
+// P[seqlen_q, seqlen_k] = Softmax(S''[seqlen_q, seqlen_k])
+// O[seqlen_q, hdim_v] = P[seqlen_q, seqlen_k] @ V^T[hdim_v, seqlen_k]
+
+namespace ck_tile {
+
+template <typename FmhaPipeline_, typename EpiloguePipeline_>
+struct FmhaFwdPagedKVKernel
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/kernel/fmha_fwd_pagedkv_kernel.hpp:27]: Add a **TODO** comment indicating that this class is a variant of the existing `FmhaFwdSplitKVKernel` pipeline. Refactoring to extract shared logic is recommended as future work.

File: example/ck_tile/01_fmha/fmha_fwd.cpp (hunk 3 above)
Comment (ltqin) [reply example/ck_tile/01_fmha/fmha_fwd.cpp:1104]: In this case, I think it can make the splitkv kernel run because if num_stiles>1, the user clearly wants to run splitkv

File: include/ck_tile/ops/fmha/pipeline/tile_fmha_traits.hpp (hunk 5 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/pipeline/tile_fmha_traits.hpp:52]: pagedkv not dropout

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp (hunk 6 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp:79]: I will remove kMergeNumHeadGroupsSeqLenQ

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp [hunk 9]
```diff
@@ -61,6 +61,59 @@ struct BlockFmhaPipelineProblem
...
+    static constexpr bool kPadSeqLenK                = Traits::kPadSeqLenK;
+    static constexpr bool kPadHeadDimQ               = Traits::kPadHeadDimQ;
+    static constexpr bool kPadHeadDimV               = Traits::kPadHeadDimV;
+    static constexpr bool kHasLogitsSoftCap          = Traits::kHasLogitsSoftCap;
+    static constexpr bool kSkipMinSeqlenQ            = Traits::kSkipMinSeqlenQ;
+    static constexpr auto BiasEnum                   = Traits::BiasEnum;
+    static constexpr bool kStoreLSE                  = Traits::kStoreLSE;
+    static constexpr bool kDoFp8StaticQuant          = Traits::kDoFp8StaticQuant;
+    static constexpr bool kIsPagedKV                 = Traits::kIsPagedKV;
+    static constexpr bool kMergeNumHeadGroupsSeqLenQ = Traits::kMergeNumHeadGroupsSeqLenQ;
```
Comment (poyenc) [inline include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp]: `kMergeNumHeadGroupsSeqLenQ` is not defined in `TileFmhaFwdPagedKVTraits<>`, please remove it.

File: include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp (hunk 4 above)
Comment (ltqin) [reply include/ck_tile/ops/gemm/block/block_gemm_areg_bsmem_creg_v2r1.hpp]: done

File: include/ck_tile/ops/fmha/pipeline/block_fmha_fwd_pagedkv_pipeline_qr_ks_vs.hpp (hunk 7 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/pipeline/block_fmha_fwd_pagedkv_pipeline_qr_ks_vs.hpp]: done

File: include/ck_tile/ops/fmha/kernel/fmha_fwd_pagedkv_kernel.hpp (hunk 8 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/kernel/fmha_fwd_pagedkv_kernel.hpp:27]: done

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp (hunk 9 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp]: done

File: example/ck_tile/01_fmha/fmha_fwd.hpp (hunk 2 above)
Comment (ltqin) [reply example/ck_tile/01_fmha/fmha_fwd.hpp:235]: want to keep all prefill function of forward

File: include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp (hunk 6 above)
Comment (ltqin) [reply include/ck_tile/ops/fmha/pipeline/block_fmha_pipeline_problem.hpp:79]: kMergeNumHeadGroupsSeqLenQ removed

**Comment by poyenc** [comment]
LGTM. We shall extract common logics from pipelines in the future.

**Review by poyenc — APPROVED** [review]
